src/locate_handle.cpp

```diff
@@ -0,0 +1,210 @@
+        }
+
+        // Extract the planar inliers from the input cloud
+        pcl::ExtractIndices<pcl::PointXYZ> extract;
+        extract.setInputCloud(cloud_cropped);
+        extract.setIndices(inliers);
+        extract.setNegative(false);
+
+        // Get the points associated with the planar surface
+        extract.filter(*cloud_plane);
+
+        // Remove the planar inliers, extract the rest
+        extract.setNegative(true);
+        extract.filter(*cloud_f);
+        *cloud_cropped = *cloud_f;
+    }
+
+    // Creating the KdTree object for the search method of the extraction
+    pcl::search::KdTree<pcl::PointXYZ>::Ptr tree(new pcl::search::KdTree<pcl::PointXYZ>);
+    tree->setInputCloud(cloud_cropped);
+
+    std::vector<pcl::PointIndices> cluster_indices;
+    pcl::EuclideanClusterExtraction<pcl::PointXYZ> ec;
+    ec.setClusterTolerance(0.01); // 1cm
+    ec.setMinClusterSize(100);
+    ec.setMaxClusterSize(25000);
+    ec.setSearchMethod(tree);
+    ec.setInputCloud(cloud_cropped);
+    ec.extract(cluster_indices);
+
+    double error = 1.0; // if the error is more than 1m, it is unlikely we found the handle
+    pcl::PointCloud<pcl::PointXYZ>::Ptr handle_cluster(new pcl::PointCloud<pcl::PointXYZ>);
+    int j = 0;
+    for (std::vector<pcl::PointIndices>::const_iterator it = cluster_indices.begin(); it != cluster_indices.end(); ++it)
+    {
+        pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_cluster(new pcl::PointCloud<pcl::PointXYZ>);
+        for (std::vector<int>::const_iterator pit = it->indices.begin(); pit != it->indices.end(); ++pit)
+            cloud_cluster->points.push_back(cloud_cropped->points[*pit]); //*
+        cloud_cluster->width = cloud_cluster->points.size();
+        cloud_cluster->height = 1;
+        cloud_cluster->is_dense = true;
+
+        Eigen::Vector4f centroid;
+        pcl::compute3DCentroid(*cloud_cluster, centroid);
+
+        // Select the cluster with centroid closest to original estimate
+        double error_x = std::abs(centroid(0) - original_handle_location.point.x);
+        double error_y = std::abs(centroid(1) - original_handle_location.point.y);
+        double error_z = std::abs(centroid(2) - original_handle_location.point.z);
+        double measured_error = error_x + error_y + error_z;
+        ROS_DEBUG_STREAM("Total error of the cluster: " << measured_error);
+        // ToDo: could also filter based on number of points to improve selection (handle ~1000 points)
+        if (measured_error < error)
+        {
+            *handle_cluster = *cloud_cluster;
+            error = measured_error;
+        }
+        ROS_DEBUG_STREAM("PointCloud representing the Cluster: " << cloud_cluster->points.size() << " data points.");
+        j++;
+    }
+
+    // check that preempt has not been requested by the client
+    if (as_->isPreemptRequested() || !ros::ok()) {
+        // set the action state to preempted
+        as_->setPreempted();
+    }
+    if (error < 1.0) {
```

Comment:
Again magic number. When changing use something like `best_fit_error` and `allowed_error`